Speedometer 3: Sync IPC in screenX / screenY

rniwa opened the pull request:
#### 3f0feeb141ba0d307884265697023af8d4b3e0a9
<pre>
Speedometer 3: Sync IPC in screenX / screenY
<a href="https://bugs.webkit.org/show_bug.cgi?id=263708">https://bugs.webkit.org/show_bug.cgi?id=263708</a>

Reviewed by Wenson Hsieh.

The sync IPC was happening because WebPageProxy::windowAndViewFramesChanged was called in
viewDidMoveToWindow before WebProcess had started and exiting early.

This resulted in WebPage to never cache the window and view coordinates, thereby forcing
WebContent process to send a send IPC to UI process whenever screenX or screenY are requested.

Fixed the bug by passing this information in WebPageCreationParameters for macOS.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/ViewWindowCoordinates.h: Added.
* Source/WebKit/Shared/ViewWindowCoordinates.serialization.in: Added.
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/Shared/WebPageCreationParameters.serialization.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::windowAndViewFramesChanged):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_historyItemClient):
(WebKit::WebPage::windowAndViewFramesChanged):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/269839@main">https://commits.webkit.org/269839@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e058e7af28c715d111d18dc94348287827a748aa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23721 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1834 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24835 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/25874 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21891 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/3425 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/24245 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/22432 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23964 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1384 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20525 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26470 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/1166 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/27684 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21638 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21714 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/25452 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/1132 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18829 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/1120 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1544 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3043 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1430 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -3179,8 +3180,10 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
     bool m_hasScheduledActivityStateUpdate { false };
     Vector<CompletionHandler<void()>> m_activityStateUpdateCallbacks;
+
+    std::unique_ptr<ViewWindowCoordinates> m_viewWindowCoordinates;
```
Should this be guarded by `PLATFORM(MAC)` instead of `PLATFORM(COCOA)`? (The other usages are all behind `PLATFORM(MAC)`).

rniwa replied:
Done.

smfr commented on Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
```diff
@@ -182,12 +182,20 @@ static inline bool expectsLegacyImplicitRubberBandControl()
+            return;
+        }
+
         send(Messages::WebPage::WindowAndViewFramesChanged(windowFrameInScreenCoordinates, windowFrameInUnflippedScreenCoordinates, viewFrameInWindowCoordinates, accessibilityViewCoordinates));
```
Why not change this message to send the new ViewWindowCoordinates struct?

rniwa replied:
Sure, we can do that.

smfr commented on Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
```diff
@@ -182,12 +182,20 @@ static inline bool expectsLegacyImplicitRubberBandControl()
     m_uiClient->windowFrame(*this, [this, protectedThis = Ref { *this }, viewFrameInWindowCoordinates, accessibilityViewCoordinates] (FloatRect windowFrameInScreenCoordinates) {
         FloatRect windowFrameInUnflippedScreenCoordinates = pageClient().convertToUserSpace(windowFrameInScreenCoordinates);
+
+        if (!hasRunningProcess()) {
```
There's some statefulness here that is tricky. We assign to `m_viewWindowCoordinates` here, but never clear it. Can we just always do this caching, and always fill `parameters.viewWindowCoordinates` on process creation? I would prefer a stateless solution.

rniwa replied:
As we discussed in the person, this is tricky to do because m_uiClient->windowFrame returns the window frame asynchronously. Will land as is. When web content process get relaunched, it would use the newly cached value, and we update the cached value whenever window moves / resizes so we should be good here.

rniwa committed:
Speedometer 3: Sync IPC in screenX / screenY
https://bugs.webkit.org/show_bug.cgi?id=263708

Reviewed by Wenson Hsieh.

The sync IPC was happening because WebPageProxy::windowAndViewFramesChanged was called in
viewDidMoveToWindow before WebProcess had started and exiting early.

This resulted in WebPage to never cache the window and view coordinates, thereby forcing
WebContent process to send a send IPC to UI process whenever screenX or screenY are requested.

Fixed the bug by passing this information in WebPageCreationParameters for macOS.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/ViewWindowCoordinates.h: Added.
* Source/WebKit/Shared/ViewWindowCoordinates.serialization.in: Added.
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/Shared/WebPageCreationParameters.serialization.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::windowAndViewFramesChanged):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_historyItemClient):
(WebKit::WebPage::windowAndViewFramesChanged):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: https://commits.webkit.org/269839@main
